### PR TITLE
[SOFT 214] data_tx solar module

### PR DESCRIPTION
**Pull request by shahhet14**
Created module that will get data from data_store and tx it into CAN.

Also includes codegen files.

### CODE REVIEW

File: projects/solar/src/test_data_tx.c
Comment (copperium) [inline projects/solar/src/test_data_tx.c:27]: I'd prefer if you didn't mock the `data_store` functions and instead used the actual data store in the test. That way we keep mocks to a minimum and the test is less flimsy (e.g. the current implementation of `data_store_get_is_set` won't work because you have to assign to `is_set` and the function should return a `StatusCode`). Then to set all the data points you can just loop through and set them all (you may want a helper function for doing this).

Also for future reference, if you use mocks in a test you have to add the test to the project's `rules.mk` file, there are examples in `solar/rules.mk`.

File: projects/solar/src/test_data_tx.c
Comment (copperium) [inline projects/solar/src/test_data_tx.c:17]: These should be static and prefixed with `s_`. I'm also not too big a fan of the comments, I'd prefer more descriptive names like e.g. `can_msg_count` instead of the comment describing what the variable does (unless it's more complex).

File: projects/solar/src/test_data_tx.c
Comment (copperium) [inline projects/solar/src/test_data_tx.c:29]: To prevent clang-format from putting your enums (or structs/unions) on one line, you can put `//` after each line, like
```c
typedef enum {
  TEST_DATA_TX_EVENT_CAN_TX = 0,  //
  TEST_DATA_TX_EVENT_CAN_RX,      //
} TestDataTxEvent;
```
`make format` will automatically align the comments for you.

Also, this should have a `TEST_DATA_TX_EVENT_CAN_FAULT` member for the fault CAN event that the settings needs, and it should end with a `NUM_TEST_DATA_TX_CAN_EVENTS` member.

File: projects/solar/src/test_data_tx.c
Comment (copperium) [inline projects/solar/src/test_data_tx.c:40]: This needs a `.fault_event = TEST_DATA_TX_EVENT_CAN_FAULT` after you've added that event. (Also, it could be a local variable in `setup_test`, but that's your call.)

File: projects/solar/src/test_data_tx.c
Comment (copperium) [inline projects/solar/src/test_data_tx.c:68]: You don't have to raise and capture the event, you can just do `Event e = { .id = DATA_READY_EVENT };` and pass that to `data_tx_process_event`.

File: projects/solar/src/test_data_tx.c
Comment (copperium) [inline projects/solar/src/test_data_tx.c:86]: Same as previous.

File: projects/solar/src/test_data_tx.c
Comment (copperium) [reply projects/solar/src/test_data_tx.c:17]: Also, `count` should be a fixed-width type, probably `uint16_t`.

File: projects/solar/src/test_data_tx.c
Comment (copperium) [inline projects/solar/src/test_data_tx.c:42]: Missing `static`.

File: projects/solar/src/test_data_tx.c
Comment (copperium) [inline projects/solar/src/test_data_tx.c:69]: Like I said before I'd prefer if you used the data store module itself instead of the mocks, so you won't have to fiddle around with `changeable_is_set`. Instead, you should give a value to all the data points here.

File: projects/solar/src/test_data_tx.c
Comment (copperium) [inline projects/solar/src/test_data_tx.c:72]: Use `uint16_t` or another fixed-width type instead of `int` (`int` and friends change size between STM32 and x86)

File: projects/solar/src/test_data_tx.c
Comment (copperium) [inline projects/solar/src/test_data_tx.c:59]: You should reset `count` in `setup_test` instead of relying on the order of test execution.

File: projects/solar/src/test_data_tx.c
Comment (copperium) [inline projects/solar/src/test_data_tx.c:88]: I'm not sure why you're checking this, didn't you just set `count` and `changeable_is_set`?

File: projects/solar/src/data_tx.c
Comment (copperium) [inline projects/solar/src/data_tx.c:25]: I might add a null check just so that we don't segfault if someone passes in `NULL`

File: projects/solar/src/test_data_tx.c
Comment (copperium) [inline projects/solar/src/test_data_tx.c:93]: Other things you could test:
* passing in `NULL` for the event
* passing in an event other than `DATA_READY_EVENT`, should do nothing
* if you change `data_tx_process_event` to return `bool`, add tests (or add to existing tests) to make sure the return value is working properly

**Comment by shahhet14** [comment]
Redesigned the data_tx module to periodically tx data using soft timers to prevent the exhaustion of the CAN queue

**Comment by shahhet14** [comment]
Redesigned the data_tx module to make the wait time between tx iterations and messages per tx iteration customizable. Data_tx now keeps tracks of messages txed using a local variable that is passed between iterations.

File: projects/solar/test/test_data_tx.c
Comment (copperium) [inline projects/solar/test/test_data_tx.c:62]: This will break if `NUM_DATA_POINTS` isn't a multiple of `msgs_per_tx_iteration`. If this ever occurs, the last `prv_process_can_events` call will try to process more TX events than there were TXes, and this will fail because `MS_TEST_HELPER_CAN_TX` will choke on the first RX event after the TX events.

To fix this, I'd set a variable at the top of the function to `MIN(s_data_tx_settings.msgs_per_tx_iteration, NUM_DATA_POINTS - s_can_msg_count)` and use that instead so that it won't try to process more TX events than are raised. (Note that the `MIN` macro is defined in misc.h.)

Even though we're keeping `msgs_per_tx_iteration` at 1 for now pending the x86 CAN bug, we'll likely increase it in the future.

File: projects/solar/test/test_data_tx.c
Comment (copperium) [inline projects/solar/test/test_data_tx.c:66]: Please use curly brackets even around single line statements!

File: projects/solar/test/test_data_tx.c
Comment (copperium) [inline projects/solar/test/test_data_tx.c:111]: Why all the casts with `int32_t`? If it's because `NUM_DATA_POINTS` is signed, I'd prefer you cast it instead, and if it's to widen everything to 32 bits I'd prefer you cast to `uint32_t` instead. Not a super big deal though.

File: projects/solar/inc/data_tx.h
Comment (copperium) [inline projects/solar/inc/data_tx.h:5]: Couple of minor typos: `Recieves => Receives`, `initiallized => initialized`.

**Review by copperium — CHANGES_REQUESTED** [review]
The only thing really blocking merging this is the point about the possible case where the test could fail. Everything else is super minor, if you address the couple of points we can get this merged!

**Comment by shahhet14** [comment]
Fixed `prv_process_can_events ` so it doesnt fail when `NUM_DATA_POINTS` is not a multiple of `msgs_per_tx_iteration`. Also fixed the other minor errors.

**Review by copperium — APPROVED** [review]

**Pull request merged**
